vertx-mongo-client/src/main/java/io/vertx/ext/mongo/impl/codec/json/JsonObjectCodec.java

```diff
@@ -39,8 +38,18 @@ public BsonValue getDocumentId(JsonObject json) {
       throw new IllegalStateException("The document does not contain an _id");
     }
 
-    String id = json.getString(ID_FIELD);
-    return new BsonString(id);
+    Object id = json.getValue(ID_FIELD);
+    if (id instanceof BsonValue) {
+      return (BsonValue) id;
+    }
+
+    BsonDocument idHoldingDocument = new BsonDocument();
```

Comment:
Given the limit set of types here (boxed primatives and JsonObject) another approach would be to explicitly construct the appropriate BsonValue sub-class for the primatives (BsonString, BsonInt64, etc...) and use BsonDocumentWrapper for JsonObject.  At a minimum it seems we should optimize for String since that's what you're going to get whenever you don't supply your own id.